0005-ordered-quantification.md

```diff
@@ -0,0 +1,626 @@
+We can augment this existing syntax with a notion of ordering, and allow quantified variables to bind to duplicate values. The key points are:
+
+* Any quantifier domain defines a potentially-infinite, partially-ordered set of *quantifier bindings*.
+* The quantified variable declarations define the values each binding maps to each variable, AND how these bindings are ordered.
+* The range expression after the `|` restricts the quantification to a subset of these bindings, but does not influence their ordering.
+
+A quantifier domain only guarantees an ordering of bindings, 
+but is NOT prescriptive on how to enumerate this domain at runtime, if it is compiled!
+This is consistent with existing expressions such as `set x: real | x in myFiniteSet && P(x)`: 
+although the unfiltered domain of `real` values is not enumerable, the bound `x in myFiniteSet` is, 
+and at runtime this comprehension is calculated by enumerating the contents of `myFiniteSet` 
+and filtering out values that do not satisfy `P(x)`. 
+The new features that are affected by quantification ordering will behave similarly: 
+`seq x: real | x in mySet && P(x)` is calculated via the same filtered enumeration, 
+but collected into a sorted sequence instead.
+
+There will be three supported variable declaration styles, and others could potentially be added in the future:
+
+1. `x [: T]`
+
+    In the existing syntax for quantifier variables (where the type `T` may be explicitly provided or omitted and inferred),
+    the bindings will be ordered according to the *natural ordering* of the type `T`. Not all Dafny types will
+    have such a natural ordering defined, but at a minimum `int` and `real`, and any subset type or newtype based on them,
+    will use the natural mathematical ordering.
+    `x: int | x in {2, 5, -4}`, for example, would bind `x` to `-4`, `2`, and `5` in that order.
+
+2. `x [: T] <- C` 
+
+    In this new syntax, the bindings will be defined and ordered according to the expression `C`, which must be a collection. Initially only the builtin collection types (`[i]set`, `multiset`, `[i]map`, and `seq`) will be supported, but support for user-defined collection types will be possible in the future. If `C` is a `map`, the bound values will be the keys of the `map`, in order to be consistent with the meaning of `x in C`; `map.Items` should be used instead to bind key-value pairs. 
+
+    Unlike the first case, this syntax may produce duplicate bindings. 
+    The ordering of bindings is non-deterministic unless `C` is a sequence.
+    If `C` is a `multiset`, multiple bindings with the same value of `x` will be created, but their ordering is still non-deterministic.
+
+    The `<-` symbol would be read as "from", so a statement like `foreach x <- C { ... }` would be read as "for each x from C, ...". Note that `<-` is not an independent operator and is intentionally distinct from the `in` boolean operator.
```

Comment:
recycling `in` might be fine?

Reply:
My concern is only that the difference between `x in C` as a quantified variable declaration and `x in C` as a boolean expression will be pretty subtle, when C is a `multiset` or `seq`: if C is `[1, 1, 1]`, for example, we'd have:

```
assert (seq x in C) == [1, 1, 1]
assert (seq x | x in C) == [1]
```

That may be less of a concern if I drop the natural ordering idea that allows the second form, so we'll see how this plays out in the next revision. :)